apps/aecore/src/aec_consensus_hc.erl

```diff
@@ -292,40 +294,84 @@ state_pre_transform_node(Type, Height, PrevNode, Trees) ->
+                    timer:sleep(child_block_time() div 4),
+                    Trees
             end;
         micro ->
+            {ok, Leader} = leader_for_timeslot(Timeslot, {TxEnv, Trees}),
             step_micro(TxEnv, Trees, Leader)
     end.
 
+step_key(Height, Timeslot, #{epoch := Epoch, first := EpochFirst, last := EpochLast} = EpochInfo,
+         {TxEnv, Trees}, Leader) ->
+    EpochFirstNonGenesis = erlang:min(?FIRST_TIMESTAMPED_BLOCK, EpochFirst),
+
+    %% Ignore actual chain height, some blocks may be missing, check whether it is the time for a new epoch
+    if Timeslot =:= EpochFirstNonGenesis ->
+        %% cache the current epoch start time
+        EpochStartTime = aetx_env:time_in_msecs(TxEnv),
+        lager:debug("[xx]STATE_PRE_TRANSFORM_NODE: CACHING EPOCH START: ~w AT HEIGHT ~w\n",[EpochStartTime, Height]),
+        cache_child_epoch_info(Epoch, Height, EpochStartTime);
+    true ->
+        ok
+    end,
+    %% note that EpochFirst and EpochLast could be the same, not exclusive
+    if Height =:= EpochLast ->
+        {Trees1, CarryOverFlag, Events} = handle_pinning(TxEnv, Trees, EpochInfo, Leader),
+        case get_entropy_hash(Epoch + 2) of
+            {ok, Seed} ->
+                lager:debug("End of epoch=~w, calling step_eoe()", [Epoch]),
+                cache_validators_for_epoch({TxEnv, Trees1}, Seed, Epoch + 2),
+                Trees2 = step_eoe(TxEnv, Trees1, Leader, Seed, 0, -1, CarryOverFlag),
+                {ok, NextEpochInfo} = aec_chain_hc:epoch_info({TxEnv, Trees2}),
+                start_default_pinning_process(TxEnv, Trees2, Height, NextEpochInfo),
+                {Trees2, Events ++ [{new_epoch, NextEpochInfo}]};
+            {error, _} ->
+                lager:debug("Entropy hash for height ~p is not in cache, attempting to resync", [Height]),
+                %% Fail the keyblock production flow, attempt to resync
+                aec_conductor:throw_error(parent_chain_not_synced)
+        end;
+    true ->
+        step(TxEnv, Trees, Leader)
+    end.
+
+start_default_pinning_process(TxEnv, Trees, _Height, NextEpochInfo) ->
+    case default_pinning_behavior() of
+        true ->
+            #{ epoch      := Epoch
+             , last       := Last
+             , validators := _Validators} = NextEpochInfo,
+            {ok, LastLeader} = leader_for_timeslot(Last, {TxEnv, Trees}),
+            lager:debug("AGENT: Trying to start pinning agent... for:  ~p in epoch ~p", [LastLeader, Epoch]),
+            try
+            case aec_parent_connector:has_parent_account(LastLeader) of
+                true -> aec_pinning_agent:spawn_for_epoch(NextEpochInfo, get_contract_pubkey(?ELECTION_CONTRACT), LastLeader);
+                false -> lager:debug("AGENT: No parent chain account found for ~p", [LastLeader])
+            end
+            catch
+                T:E -> lager:debug("AGENT throws: ~p:~p", [T,E])
+            end;
+        _ -> ok
+    end.
```

Comment:
Old stuff, right?

Reply:
Will try and roll it back, these changes do not belong here but got them from Richard's branch which is somewhat old.